### ISSUE TITLE
Converting 8 bit to 1 bit mask data

### ISSUE THREAD

**Issue by blasky** [kleisauke/net-vips#115]
I've been working on converting some imagery and metadata to the OME-TIFF format. Most things have been going well. But I'm a little stuck on a conversion I need to do for their masking data.

Masks in my system are stored as 1 band, 8 bit, b-w, Png files (...and also a 4 band, rgba version for further display purposes). I believe the mask data that OME-TIFF requires is 1 bit, unformatted, pixel data. (And then the data will be Base64 encoded before being included in the Xml metadata.)

I've been looking at Image.Ifthenelse() to set a threshold for 0/1 - black/white data. And I was looking at Image.Cast() to see if there was a BandFormat that could get me to 1 bit. But I've been coming up empty so far.

Is there a straight forward NetVips way to make this conversion/translation?

(The mask files are very small to start with so ideally I will do this in memory. But I could write out  a temp file if that made a difference.)

Thanks.

**Comment by blasky** [comment]
I do see that I can get to 1 band via Tiff like so...
```
using (Image maskPng = Image.Pngload(maskFile))
{
   byte[] maskBytes = maskPng.TiffsaveBuffer(strip: true, bitdepth: 1);
   string maskStrData = Convert.ToBase64String(maskBytes);
}
```
And I included the strip: parameter to remove any Tiff metadata. But this is still Tiff formatted and OME doesn't seem to like it. Is there a Raw pixel equivalent to this?

**Comment by blasky** [comment]
This approach "works". Using the NetVips extension to convert to Bitmap and GetPixel calls. But it is obviously very inefficient and brute force like. I'm sure there is still a better way.
```
using (Image mask8Bit = Image.Pngload(maskFile))
using (Bitmap maskBitmap = mask8Bit.ToBitmap())
{
   byte[] maskRaw = new byte[maskBitmap.Width * maskBitmap.Height];

   for (int y = 0; y < maskBitmap.Height; y++)
   {
      for (int x = 0; x < maskBitmap.Width; x++)
      {
         maskRaw[(y * maskBitmap.Width) + x] = (byte)(maskBitmap.GetPixel(x, y).R > 0 ? 1 : 0);
      }
   }

   maskStrData = Convert.ToBase64String(maskRaw);
}
```

**Comment by kleisauke** [comment]
Hi @blasky,

I think the above pixel loop can be written as:
```csharp
using var black = Image.Black(10, 10, bands: 3);
using var rgb = black.Embed(0, 5, 10, 10, extend: Enums.Extend.White);
using var r = rgb[0];
using var im = r.Ifthenelse(1, 0);
var maskBinary = im.WriteToMemory();

// 1, 1, 1, 1, 1, 1, 1, 1, 1, 1,
// 1, 1, 1, 1, 1, 1, 1, 1, 1, 1,
// 1, 1, 1, 1, 1, 1, 1, 1, 1, 1,
// 1, 1, 1, 1, 1, 1, 1, 1, 1, 1,
// 1, 1, 1, 1, 1, 1, 1, 1, 1, 1,
// 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
// 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
// 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
// 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
// 0, 0, 0, 0, 0, 0, 0, 0, 0, 0
Console.WriteLine(string.Join(", ", maskBinary));
```

/cc @jcupitt Do you think a `bitdepth` parameter for `vips_image_write_to_memory` makes sense?

**Comment by blasky** [comment]
Of course. Close to that when I was messing with Ifthenelse() but not exactly. Thanks.
This seems totally reasonable to me, even without a one line WriteToMemory(bitdepth: 1) solution.
```
using (NetVips.Image mask8Bit = NetVips.Image.Pngload(maskFile)[0])
using (NetVips.Image mask1Bit = mask8Bit.Ifthenelse(1, 0))
{
   byte[] maskBinary = mask1Bit.WriteToMemory();
   maskStrData = Convert.ToBase64String(maskBinary);
}
```

**Comment by jcupitt** [comment]
You can do it without the `ifthenelse`: test for `!=0` to get 255 (true) for non-zero, then take the bottom bit with `&1`. In Python (sorry, I don't know C#):

```
bit_mask = (image[0] != 0) & 1
```

I'm not sure about a bitdepth param for `write_to_memory`. Every image format that supports 1 bit images seems to have slightly different rules about ordering and alignment (most-significant bit first / least-significant bit first, pad line endings to byte boundary / pad to 32-bit word boundary, etc.). Perhaps it's better linked to specific image formats?

The libvips tiff load and save supports OME-TIFF pyramids, you probably saw.

**Comment by kleisauke** [comment]
Thanks! I updated the above C# example. Indeed, a bitdepth parameter for `write_to_memory` is complicated to do due to different ordering/aligning rules (I hadn't thought of that for a moment).
